File: app/app/src/test/java/dev/steamvault/app/ui/detail/logic/DepotPresentationTest.kt
package dev.steamvault.app.ui.detail.logic

import dev.steamvault.app.net.model.GameSummary
import dev.steamvault.app.ui.library.logic.MultiPlanDepotRow
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pins the four depot-sharing wordings [DepotShareTag] can produce
 * (WP 4b.6 brief: "mutation-worthy pins (sharing wording per case)"), and
 * the co-owner row cached/not-cached split feeding the mockup's verbatim
 * "not cached · mapping kept" state (mockup-notes.md round 3).
 */
class DepotPresentationTest {

    private val other1 = GameSummary(201, "Other One", "done", "2026-08-01T00:00:00Z", null, 1, 1_000_000_000, false)
    private val other2 = GameSummary(202, "Other Two", "idle", null, null, 0, null, false)
    private val gamesByAppid = mapOf(other1.appid to other1, other2.appid to other2)

    @Test
    fun `not shared -- EXCLUSIVE, no tag, no co-owners`() {
        val row = MultiPlanDepotRow(depotid = 1, sizeBytes = 500L, shared = false, others = emptyList(), holderAppids = emptyList(), free = true)
        val presentation = buildDepotPresentation(row, gamesByAppid, thisAppIsHolder = true)
        assertEquals(DepotShareTag.EXCLUSIVE, presentation.tag)
        assertTrue(presentation.coOwners.isEmpty())
    }

    @Test
    fun `shared with a cached co-owner -- PROTECTED regardless of whether this app is a holder`() {
        val row = MultiPlanDepotRow(
            depotid = 2, sizeBytes = 500L, shared = true,
            others = listOf(other1.appid), holderAppids = listOf(other1.appid), free = false,
        )
        val presentation = buildDepotPresentation(row, gamesByAppid, thisAppIsHolder = true)
        assertEquals(DepotShareTag.PROTECTED, presentation.tag)
        assertEquals(listOf(CoOwnerRow(other1.appid, "Other One", cached = true)), presentation.coOwners)
    }

    @Test
    fun `shared, no other holder, THIS app holds it -- SOLE_HOLDER (mockup round 5)`() {
        val row = MultiPlanDepotRow(
            depotid = 3, sizeBytes = 500L, shared = true,
            others = listOf(other2.appid), holderAppids = emptyList(), free = true,
        )
        val presentation = buildDepotPresentation(row, gamesByAppid, thisAppIsHolder = true)
        assertEquals(DepotShareTag.SOLE_HOLDER, presentation.tag)
        assertEquals(listOf(CoOwnerRow(other2.appid, "Other Two", cached = false)), presentation.coOwners)
    }

    @Test
    fun `shared, no other holder, THIS app also does not hold it -- ORPHANED (mapping kept, nobody cached)`() {
        // Mirrors mockup-notes.md's Meridian Rally sample: a previously
        // deleted game whose shared depot's remaining co-owners are ALSO
        // all uncached.
        val row = MultiPlanDepotRow(
            depotid = 4, sizeBytes = 500L, shared = true,
            others = listOf(other2.appid), holderAppids = emptyList(), free = true,
        )
        val presentation = buildDepotPresentation(row, gamesByAppid, thisAppIsHolder = false)
        assertEquals(DepotShareTag.ORPHANED, presentation.tag)
    }

    @Test
    fun `an unresolvable co-owner falls back to the App id name, mirrors GameCardModel and JobCardModel`() {
        val row = MultiPlanDepotRow(
            depotid = 5, sizeBytes = null, shared = true,
            others = listOf(999), holderAppids = emptyList(), free = true,
        )
        val presentation = buildDepotPresentation(row, emptyMap(), thisAppIsHolder = true)
        assertEquals("App 999", presentation.coOwners.single().name)
    }

    @Test
    fun `MUTATION TARGET -- a holder appid is reported cached, a non-holder co-owner is not`() {
        val row = MultiPlanDepotRow(
            depotid = 6, sizeBytes = 500L, shared = true,
            others = listOf(other1.appid, other2.appid), holderAppids = listOf(other1.appid), free = false,
        )
        val presentation = buildDepotPresentation(row, gamesByAppid, thisAppIsHolder = false)
        val byAppid = presentation.coOwners.associateBy { it.appid }
        assertTrue(byAppid.getValue(other1.appid).cached)
        assertTrue(!byAppid.getValue(other2.appid).cached)
    }
}

File: web/tests/depot-presentation.test.js
/**
 * Headless tests for web/js/lib/depot-presentation.js (WP 4a.4).
 *
 * Row shape mirrors lib/multiplan.js's `MultiPlanDepotRow`
 * ({depotid, sizeBytes, shared, others, holderAppids, free}) for a
 * single-id batch. Ported alongside the Android sibling's
 * `DepotPresentationTest.kt` (WP 4b.6) — same four states, same reasoning.
 *
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { buildDepotPresentation, DEPOT_TAG } from "../js/lib/depot-presentation.js";

function row(overrides) {
  return { depotid: 441, sizeBytes: 1_000_000, shared: false, others: [], holderAppids: [], free: true, ...overrides };
}

test("EXCLUSIVE: an unshared depot never gets a tag, regardless of thisAppIsHolder", () => {
  const gamesByAppid = new Map();
  assert.equal(buildDepotPresentation(row({ shared: false }), gamesByAppid, true).tag, DEPOT_TAG.EXCLUSIVE);
  assert.equal(buildDepotPresentation(row({ shared: false }), gamesByAppid, false).tag, DEPOT_TAG.EXCLUSIVE);
});

test("EXCLUSIVE: has no co-owners (multiplan.js never populates `others` for an unshared row)", () => {
  const p = buildDepotPresentation(row({ shared: false, others: [] }), new Map(), true);
  assert.deepEqual(p.coOwners, []);
});

test("PROTECTED: a shared depot with at least one OTHER cached holder is PROTECTED regardless of thisAppIsHolder", () => {
  const gamesByAppid = new Map([[730, { name: "Other Cached" }]]);
  const r = row({ shared: true, others: [730], holderAppids: [730], free: false });
  assert.equal(buildDepotPresentation(r, gamesByAppid, true).tag, DEPOT_TAG.PROTECTED);
  assert.equal(buildDepotPresentation(r, gamesByAppid, false).tag, DEPOT_TAG.PROTECTED);
});

test("SOLE_HOLDER: a shared depot with no OTHER cached holder, and THIS app holds it", () => {
  const gamesByAppid = new Map([[730, { name: "Other Uncached" }]]);
  const r = row({ shared: true, others: [730], holderAppids: [], free: true });
  assert.equal(buildDepotPresentation(r, gamesByAppid, true).tag, DEPOT_TAG.SOLE_HOLDER);
});

test("ORPHANED (recorded WP 4b.6 divergence, adopted here): a shared depot with no cached holder at all, incl. THIS app", () => {
  // The last-remnant/Meridian-Rally scenario (ADR-0003): the VIEWED game
  // does not hold this depot either -- neither it nor any co-owner is
  // cached. Collapsing this into SOLE_HOLDER would state a falsehood.
  const gamesByAppid = new Map([[730, { name: "Other Uncached" }]]);
  const r = row({ shared: true, others: [730], holderAppids: [], free: true });
  assert.equal(buildDepotPresentation(r, gamesByAppid, false).tag, DEPOT_TAG.ORPHANED);
});

test("coOwners: resolves names from gamesByAppid, falls back to 'App {appid}' for an unknown one", () => {
  const gamesByAppid = new Map([[730, { name: "Known Game" }]]);
  const r = row({ shared: true, others: [730, 999], holderAppids: [730] });
  const p = buildDepotPresentation(r, gamesByAppid, false);
  assert.deepEqual(
    p.coOwners.map((c) => ({ appid: c.appid, name: c.name, cached: c.cached })),
    [
      { appid: 730, name: "Known Game", cached: true },
      { appid: 999, name: "App 999", cached: false },
    ],
  );
});

test("passes depotid and sizeBytes through unchanged", () => {
  const p = buildDepotPresentation(row({ depotid: 12345, sizeBytes: 42 }), new Map(), false);
  assert.equal(p.depotid, 12345);
  assert.equal(p.sizeBytes, 42);
});
